crates/bevy_ecs/src/entity/remote_allocator.rs

```diff
@@ -0,0 +1,1052 @@
+        self.shared.free.num_free()
+    }
+
+    /// Frees the entity allowing it to be reused.
+    #[inline]
+    pub(super) fn free(&mut self, entity: Entity) {
+        // SAFETY: We have `&mut self`.
+        unsafe {
+            self.shared.free.free(entity);
+        }
+    }
+
+    /// Allocates `count` entities in an iterator.
+    #[inline]
+    pub(super) fn alloc_many(&self, count: u32) -> AllocEntitiesIterator<'_> {
+        // SAFETY: `free` takes `&mut self`, and this lifetime is captured by the iterator.
+        unsafe { self.shared.alloc_many(count) }
+    }
+}
+
+impl Drop for Allocator {
+    fn drop(&mut self) {
+        self.shared.close();
+    }
+}
+
+impl core::fmt::Debug for Allocator {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct(core::any::type_name::<Self>())
+            .field("total_indices", &self.total_entity_indices())
+            .field("total_free", &self.num_free())
+            .finish()
+    }
+}
+
+/// An [`Iterator`] returning a sequence of [`Entity`] values from an [`Allocator`].
+///
+/// **NOTE:** Dropping will leak the remaining entities!
+pub(super) struct AllocEntitiesIterator<'a> {
+    new: AllocUniqueEntityIndexIterator,
+    reused: FreeBufferIterator<'a>,
+}
+
+impl<'a> Iterator for AllocEntitiesIterator<'a> {
+    type Item = Entity;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.reused.next().or_else(|| self.new.next())
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.reused.len() + self.new.len();
+        (len, Some(len))
+    }
+}
+
+impl<'a> ExactSizeIterator for AllocEntitiesIterator<'a> {}
+impl<'a> core::iter::FusedIterator for AllocEntitiesIterator<'a> {}
+
+// SAFETY: Newly reserved entity values are unique.
+unsafe impl EntitySetIterator for AllocEntitiesIterator<'_> {}
+
+impl Drop for AllocEntitiesIterator<'_> {
+    fn drop(&mut self) {
+        let leaking = self.len();
+        if leaking > 0 {
+            warn!(
+                "{} entities being leaked via unfinished `AllocEntitiesIterator`",
+                leaking
+            );
+        }
+    }
+}
+
+/// This is a stripped down entity allocator that operates on fewer assumptions than [`EntityAllocator`](super::EntityAllocator).
+/// As a result, using this will be slower than than the main allocator but this offers additional freedoms.
+/// In particular, this type is fully owned, allowing you to allocate entities for a world without locking or holding reference to the world.
+/// This is especially useful in async contexts.
+#[derive(Clone)]
+pub struct RemoteAllocator {
+    shared: Arc<SharedAllocator>,
+}
+
+impl RemoteAllocator {
+    /// Creates a new [`RemoteAllocator`] with the provided [`Allocator`] source.
+    /// If the source is ever destroyed, [`Self::alloc`] will yield garbage values.
+    /// Be sure to use [`Self::is_closed`] to determine if it is safe to use these entities.
```

Comment:
This should probably be [`EntityAllocator::has_remote_allocator`](super::EntityAllocator::has_remote_allocator) now? Maybe `is_closed()` can be removed completely, to prevent misuse? I guess if it's a diagnostic thing, it could be called `weak_is_closed()` to make sure it's a "weak" check (like a weak pointer)?